### DISCUSSION TITLE
[trace] Error while reading build.prop file:Error #3001: File or directory access denied.

### DISCUSSION THREAD

**Discussion by phitrti** [airsdk/Adobe-Runtime-Support#2494] [Q&A]
I've read a bunch of information about Android file permissions and 3001 errors. A tad overwhelming to be honest. I think what I'm seeing is slightly different from other issues and i'm not really sure what is going on.

I'm debugging my app on Android Samsung tablet from IntelliJ with AIR 50.2.1 using a MacBook running Monterey. While my splash screen is displayed and before I land in my entry point for the app, I'm getting:

[trace] Error while reading build.prop file:Error #3001: File or directory access denied.

I would love to know what this is trying to access. Seems to be a startup item but the only reference i found to build.prop was something about Adobe Muse in 2019 and MacOS file permissions. 

My app runs fine so far if i just pretend that error doesn't happen. I use applicationstoragedirectory and the app "appears" to be working. It has been a while since I debugged on Android as most of my air work the last few years was iOS. Am I missing something at startup?  I have the following in my app xml file:

			    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
			    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>

and I try to avoid using documentsdirectory.

Thoughts? Even though the app seems to work, i hate dangling errors like that because something really could be a problem internally somewhere.

Randy

**Comment by teotigraphix** [comment]
Check out the readme.pdf for AIR 50.2.1. Andrew explains the file system changes and the File API for Android.

Bottom line is, nothing is accessible without permissions(except local process/user storage) and the readme explains how it's different.

Not sure about the build.prop but yeah something is trying to access a non internal file.

**Reply by phitrti** [reply]
Right, Android has really locked things down and the readme is most helpful. But Build.prop appears to be an android system file. I'm not trying to access it directly so I'm unclear why that error is being thrown. I'm searching for any static variables that might reference the file system (there shouldn't be).

**Comment by reckhoff** [comment]
Great question. Appears to be AIR. I use IntelliJ IDEA as my IDE:

"/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home/bin/java" -Dapplication.home=/opt/tools_air/flexsdks/4.16.1_AIR50.2.1 -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -classpath /opt/tools_air/flexsdks/4.16.1_AIR50.2.1/lib/fdb.jar flex.tools.debugger.cli.DebugCLI -ide -p 7936
Apache fdb (Flash Player Debugger) [build 20171115]
Copyright 2017 The Apache Software Foundation. All rights reserved.
Trying to connect to Player
Player connected; session starting.
[trace] Error while reading build.prop file:Error #3001: File or directory access denied.

-Randy

From: Michael Schmalle ***@***.***>
Date: Monday, February 27, 2023 at 4:16 PM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] [trace] Error while reading build.prop file:Error #3001: File or directory access denied. (Discussion #2494)
EXTERNAL:  This email originated from outside of the organization.  Do not click links or open attachments unless you recognize the sender and know the content is safe.


unclear why that error is being thrown.

Where is this error thrown, AIR or Android OS?

**Reply by ajwfrost** [reply]
Curious as to whether this happens without Flex too..? I can't see anything in the AIR runtime that would try to open build.prop - although there is a comment that the "DensityDpi" API gets its information from the device's build.prop file.

What device are you running on?

I've just had a quick try and I don't get that message either on the application start-up (and I've got a debug build which traces out all filesystem requests) or when I manually call that DisplayMetrics.densityDpi value.. so either I'm not using the right kind of handset/version, or it's a flex thing (though I couldn't find this mentioned in the Flex AS3 source code..)

thanks

**Comment by phitrti** [comment]
This is a flex app and I’m debugging on a Samsung Tablet Tab S6 Lite running Android 13.

Today is a crazy day but will try to investigate more on my side’s startup code.

-Randy

From: Andrew Frost ***@***.***>
Date: Monday, February 27, 2023 at 6:31 PM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: [RCD] PHIT ***@***.***>, Author ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] [trace] Error while reading build.prop file:Error #3001: File or directory access denied. (Discussion #2494)
EXTERNAL:  This email originated from outside of the organization.  Do not click links or open attachments unless you recognize the sender and know the content is safe.


Curious as to whether this happens without Flex too..? I can't see anything in the AIR runtime that would try to open build.prop - although there is a comment that the "DensityDpi" API gets its information from the device's build.prop file.

What device are you running on?

I've just had a quick try and I don't get that message either on the application start-up (and I've got a debug build which traces out all filesystem requests) or when I manually call that DisplayMetrics.densityDpi value.. so either I'm not using the right kind of handset/version, or it's a flex thing (though I couldn't find this mentioned in the Flex AS3 source code..)

thanks

**Comment by reckhoff** [comment]
Your DPI comment got me thinking…. So I did a grep on “build.prop” in the Flex source code.

Frameworks/projects/mobilecomponents/src/spark/utils/PlatformMobileHelper.as has a function that reads from build.prop. Line 64 is the trace statement of the message I see. So grepping backwards, “computeOSVersionForAndroid” is used by frameworks/projects/framework/src/mx/utils.Platform.as osVersion(). I guess this could be called by a lot of places but grepping on “Platform.osVersion” as it is a static function points me to frameworks/projects/framework/src/mx/utils/MediaQueryParser.as getOSVersion function. It is a private function that gets called by the MediaQueryParser constructor. And this is where a lot of DPI is determined. I don’t see the osVersion used to help with the determination of the dpi, just that it is one of the properties saved in this class. Maybe other flex classes call Platform.osVersion via reflection. I did not grep on just “osVersion”.

Moving off of flex isn’t an option. I’m currently searching for Android options, maybe a permission that needs to be set in the manifest, that would allow access to that file. If you have any other ideas…..

-Randy

From: phitrti ***@***.***>
Date: Tuesday, February 28, 2023 at 7:12 AM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: Eckhoff, Randy ***@***.***>, Comment ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] [trace] Error while reading build.prop file:Error #3001: File or directory access denied. (Discussion #2494)
EXTERNAL:  This email originated from outside of the organization.  Do not click links or open attachments unless you recognize the sender and know the content is safe.

This is a flex app and I’m debugging on a Samsung Tablet Tab S6 Lite running Android 13.

Today is a crazy day but will try to investigate more on my side’s startup code.

-Randy

From: Andrew Frost ***@***.***>
Date: Monday, February 27, 2023 at 6:31 PM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: [RCD] PHIT ***@***.***>, Author ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] [trace] Error while reading build.prop file:Error #3001: File or directory access denied. (Discussion #2494)
EXTERNAL: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


Curious as to whether this happens without Flex too..? I can't see anything in the AIR runtime that would try to open build.prop - although there is a comment that the "DensityDpi" API gets its information from the device's build.prop file.

What device are you running on?

I've just had a quick try and I don't get that message either on the application start-up (and I've got a debug build which traces out all filesystem requests) or when I manually call that DisplayMetrics.densityDpi value.. so either I'm not using the right kind of handset/version, or it's a flex thing (though I couldn't find this mentioned in the Flex AS3 source code..)

thanks

**Reply by teotigraphix** [reply]
>  maybe a permission that needs to be set in the manifest

The only way to solve this is finding the call. Since Android has the sandboxed user data now with public swimming pool, anything you do to get into the main pool is a permission. The real problem here is you don't even have access to the file stream call happening.

To even get a permission now in Android, you have to get a handle to the file/directory and THEN ask for the permission on that file handle,. which you don't have. There is no blanket permissions up front or even asking, you need to know where to ask.

I am just trying to save you time here. ;-)

**Reply by ajwfrost** [reply]
Well found @reckhoff ! but it sounds like any workaround may require an update in that Flex class. (I think - not 100% sure but it may be possible - you could actually define a class "spark.utils.PlatformMobileHelper" within your own application, and then this class definition would be used in preference to the one from the Flex library at runtime?)

There's potential mention of a workaround here:
https://github.com/Fazecast/jSerialComm/issues/122
but not quite sure how this `getprop` thing is used or could be called, maybe an ANE is required here!

**Comment by phitrti** [comment]
Thanks @teotigraphix ! yea, no blanket permission for system files. Can you imagine the user experience at startup "we want to access the system file so we can determine the android version number. do you grant us permission to do that?". gads, what a UX. Especially in a class I have no control over and is in an SDK that others use. I think Android went too far in their file permissions.

@ajwfrost  - brilliant! Replace PlatformMobileHelper with my own. I did exactly that and the error goes away. I don't know if the following is the "right" solution for os version or how Distriqt's Application ANE determines "device.os.version" but the following is my implementation of PlatformMobileHelper. Works right now and I don't know if something is broken downstream. Maybe this will help other flex users down the road. I put this class in my main application so it is found first before the Flex provided version.

`package spark.utils {
import com.distriqt.extension.application.Application;
public class PlatformMobileHelper {


    public static function computeOSVersionForAndroid(): String
    {
        var version: String = "";
        version = Application.service.device.os.version;

        return version;
    }

}
}`

**Reply by teotigraphix** [reply]
Well, there is a "blanket" but we don't get it. :) This is an interesting read to get up to speed;

https://developer.android.com/training/data-storage/manage-all-files
